### PR TITLE
feat(CFD): Add binary option CFD implementation and tests

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

As part of our set of financial contracts built on the CFD we require a simple binary option contract library. This PR adds this implementation and tests.

### CODE REVIEW

File: packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/BinaryOptionContractForDifferenceFinancialProductLibrary.sol [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+
+import "./ContractForDifferenceFinancialProductLibrary.sol";
+import "../../../../common/implementation/Lockable.sol";
+
+/**
+ * @title Binary Option Contract For Difference Financial Product Library.
+ * @notice Adds settlement logic to simply binary option CFDs. Binary options settle with all collateral allocated to
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/BinaryOptionContractForDifferenceFinancialProductLibrary.sol:11]:
```suggestion
 * @notice Adds settlement logic to binary option CFDs. Binary options settle with all collateral allocated to
```

File: packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/BinaryOptionContractForDifferenceFinancialProductLibrary.sol [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * settlement price is greater or equal to the strike then all value is sent to the long side. Otherwise, all value
+ * is sent to the short side. The settlement price could be a scalar (like the price of ETH) or a binary bet with
+ * settlement being 0 or 1 depending on the outcome.
+ */
+contract BinaryOptionContractForDifferenceFinancialProductLibrary is
+    ContractForDifferenceFinancialProductLibrary,
+    Lockable
+{
+    using FixedPoint for FixedPoint.Unsigned;
+    using SafeMath for uint256;
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/BinaryOptionContractForDifferenceFinancialProductLibrary.sol:23]: is safemath needed in 0.8?

File: packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/BinaryOptionContractForDifferenceFinancialProductLibrary.sol [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     * c) A strike price can only be set once to prevent the deployer from changing the strike after the fact.
+     * d) For safety, a strike price should be set before depositing any synthetic tokens in a liquidity pool.
+     * e) financialProduct must expose an expirationTimestamp method to validate it is correctly deployed.
+     */
+    function setContractForDifferenceParameters(address contractForDifference, int256 strikePrice)
+        public
+        nonReentrant()
+    {
+        require(ExpiringContractInterface(contractForDifference).expirationTimestamp() != 0, "Invalid CFD address");
+        require(contractForDifferenceParameters[contractForDifference].isSet == false, "Parameters already set");
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/BinaryOptionContractForDifferenceFinancialProductLibrary.sol:46]:
```suggestion
        require(!contractForDifferenceParameters[contractForDifference].isSet, "Parameters already set");
```

**Review by mrice32 — APPROVED** [review]
LGTM!

File: packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/BinaryOptionContractForDifferenceFinancialProductLibrary.sol (hunk 2 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/BinaryOptionContractForDifferenceFinancialProductLibrary.sol:23]: no but we import it everywhere else! I don't think it adds much gas as the OZ lib simply does unchecked operations under the hood and relies on the solidity validation

**Pull request merged**